### ISSUE TITLE
Support standard .NET Core DI

### ISSUE THREAD

**Issue by ramondeklein** [shinyorg/shiny#545]
## Summary
Shiny uses its own mechanism of registering components for DI, which makes it difficult to integrate Shiny in other applications. I am trying to create an ASP.NET web API, but it's hard to combine ASP.NET startup mechanism with Shiny's DI.

## API Changes
I would suggest to replace the `ShinyHost.Init` with two separate calls:
* `ShinyHost.RegisterPlatform(this IServiceCollection services, IPlatform platform)` that registers the current platform (calls the `AddSingleton` and `IPlatform.Register` with the service collection.
* `ShinyHost.RegisterContainer(IServiceProvider serviceProvider)` that saves the actual DI root-container.

Of course the current `ShinyHost.Init` can be kept to maintain backward compatibility or as a convenience wrapper for people that prefer the Shiny way of DI.

It is already possible to register the platform by duplicating some code from `ShinyHost.Init`, but I can't assign the root DI-container, because the `container` field is a private static that is set via `ShinyHost.Init`. I do think, I could hack around and create another root DI container (via `Startup.CreateServiceProvider`), Besides, very hacky, this would expose the incorrect service collection (via `ShinyHost.Services`).

## Intended Use Case
It would be much easier to integratie Shiny into DI systems that create the `ServiceCollection` object (just like Shiny does). In the ASP.NET Core case, the `Startup` class can call `RegisterPlatform` in its `ConfigureServices` method and when the actual root container is created (typically after a HostBuilder.Build() call), then it's handed over to Shiny.

**Comment by aritchie** [comment]
Shiny doesn't support ASP.NET so I'm not even sure why you are trying to integrate with that.  

This will be changing in a future release to support the hostbuilder to integrate properly with MAUI, but that is 6 months out

**Comment by ramondeklein** [comment]
I integrate with ASP.NET, because I needed it :-) My Xamarin app exposes a web API (localhost only) and provides a web-browser control that is loaded with an Angular app. This app is hosted from embedded resources and it works quite well. We can use the same front-end developers as for our other web apps and I don't need to invest in a (a soon to be deprecated) Xamarin Forms app.

Kestrel is a pain with Xamarin, but I managed to get ASP.NET Core v2.2.0 working with the latest Xamarin version and Shiny. The main problem is that ASP.NET is integrated with later .NET Core frameworks, but not available for Xamarin and there are no NuGet packages anymore. I hope that -with .NET Core 6- it will converge to a single eco-system and we can also host an ASP.NET server from a Xamarin app.

The Kestrel/Xamarin issue has nothing to do with Shiny, but is just a sample use-case. I can imagine more use-cases that use the hostbuilder mechanism. I now use a kind of interop-layer to deal with ASP.NET DI and Shiny DI. It works, but it would be more elegant to use a single DI mechanism (that's why I created this PR).

It's great to hear that that hostbuilder support is on the roadmap, but I hope Shiny won't be tied to MAUI.  It would be great to use Shiny with plain Xamarin (although convenience methods to easily integrate with MAUI are very useful). A more loosely connection to DI registration would also allow Shiny to be used in different scenarios. But I don't know enough about Shiny's internals to know if this is possible at all.

Just my 2 cents...
